appsrc/dcfile/dccp.man
.TH DCCP 1 "6 March 2014" "DICOM PS3" "DICOM PS3 - Copy DICOM file"
.SH NAME
dccp \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Copy DICOM file
.SH SYNOPSIS
.HP 10
.B dccp
.so man1/gen.so
[
.B \-v|verbose
]
[
.B \-ignorereaderrors
]
[
.B \-removecommandgroup
]
[
.B \-fixbaddecimalseparator
]
[
.B \-selectvoilutitem 0..n\-1
]
.so man1/optin.so
.so man1/optout.so
.SH DESCRIPTION
.LP
.B dccp
reads the named dicom or acr-nema input file and copies the information and
pixel data to a new dicom file.
.LP
The input encoding of the file will be automatically determined, or can
be explicitly specified in pathological cases using the input options
described in dcintro(1).
.LP
The encoding of the file will be changed to the default or explicitly
requested transfer syntax, as described in dcintro(1) under output
options. This provides a means of converting between different transfer
syntaxes, and adding or removing Part 10 style meta-information headers
to or from "true" DICOM files, DICOM messages captured from network
exchanges such as with the Mallinckrodt or Oldenburg central test node
software, or converting old ACR/NEMA or SPI files.
.LP
During the translation, specific attributes may be deleted, added or
replaced, class and instance unique identifiers generated or removed,
group lengths added, or private attributes removed, as specified by
the replacement options described in dcintro(1). This allows pre-DICOM
ACR-NEMA or SPI files to be converted to DICOM files containing
conforming instances of SOP classes in a semi-automated manner. If patient and study information is present (indicating a composite IOD), and the
SOP Class can be automatically determined, such as for CT or MR images,
specific SOP Class instances will be created, otherwise SC will be used.
Unique study, series and instance identifiers will be generated from a
time stamp, but can be overridden by command line options to allow for
shared identifiers between files of a series.
.SH OPTIONS
The verbose output goes to standard error.
.LP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-verbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents while reading, once read, during replacement, and before writing.
.RE
.TP
.B \-ignorereaderrors
.RS
Normally dccp stops if read errors are encountered, but if this option is
turned on, it will try to write what it can. This may be useful to remove
trailing garbage after a valid dicom set that has been padded past the
pixel data (e.g. to a blocksize) or similar.
.RE
.TP
.B \-removecommandgroup
.RS
Normally dccp will created an invalid copy if the input is in implicit VR and erroneously has
command group (0x0000) elements at the head of the dataset (e.g. if the both the command
and dataset network PDUs were saved to a file), since the command group elements will get
inserted in fromt of the meta information header. Use this option to remove command group
elements (whether or not a meta information header is added).
.RE
.TP
.B \-fixbaddecimalseparator
.RS
Sometimes illegal DS VR values are encountered that use a Locale-specific decimal separator of "," (such as might be displayed in France, Italy, Brazil, etc.) rather
than the period "." required by the standard; this option can be used to fix these.
.RE
.TP
.B \-selectvoilutitem 0..n\-1
.RS
Keep only the selected VOI LUT Sequence Item, where the items are numbered from 0.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% echo Adds meta-information header by default
.RE
% dccp -v NM.dc3 NM2.dc3
.RE
******** While reading ... ********
.RE
@0x00000000,0x00000000 of 0xffffffff: (0x0008,0x0000) ...
.RE
...
.RE
******** As read ... ********
.RE
(0x0008,0x0000) UL Group Length VR=<UL> ...
.RE
...
.RE
******** As written ... ********
.RE
(0x0002,0x0000) UL Meta Element Group Length ...
.RE
(0x0002,0x0001) OB File Meta Information Version ...
.RE
...
.RE
(0x0008,0x0008) CS Image Type VR=<CS> VL=<24> <...>
.RE
...
.RE
\ 
.RE
% echo Make explicit big endian, with stamp & replacement,
.RE
       and removal of private attributes
.RE
\ 
.RE
% dccp \-v \-ts 1.2.840.10008.1.2.2 \-stamp "9999" 
.RE
       \-r PatientSex F \-removeprivate
.RE
       test.dc3 test.big
.RE
******** While reading ... ********
.RE
@0x00000000,0x00000000 of 0xffffffff: (0x0008,0x0000) ...
.RE
...
.RE
******** As read ... ********
.RE
...
.RE
(0x0010,0x0040) CS PatientSex VR=<CS> VL=<2> <M > 
.RE
...
.RE
(0x0028,0x0010) US Rows       VR=<US> VL=<2> [100]
.RE
...
.RE
(0x0029,0x0010) LO Private Creator ... [ACME CO ]
.RE
(0x0029,0x1010) DS Zoom       VR=<DS> VL=<0> <>
.RE
...
.RE
******** After replace before... ********
.RE
...
.RE
(0x0010,0x0040) CS PatientSex VR=<CS> VL=<2> <F> 
.RE
...
.RE
******** As written ... ********
.RE
(0x0002,0x0000) UL Meta Element Group Length ...
.RE
(0x0002,0x0001) OB File Meta Information Version ...
.RE
...
.RE
(0x0010,0x0040) CS PatientSex VR=<CS> VL=<2> <F> 
.RE
...
.RE
(0x0028,0x0010) US Rows       VR=<US> VL=<2> [100]
.RE
...
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dcintro(1),dccreate(1)
.SH AUTHOR
Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
241220:	libsrc/src/dctool/attrtypv.cc: Allow TAB character in control characters for text VRs (CP 1425)

241114:	appsrc/dcfile/dciodvfy.cc: Allow single byte of padding for odd number of 8 bit LUT entries in 16 bit words (000607)

241114:	appsrc/dcfile/dciodvfy.cc: Remove check on SegmentNumber monotonically increasing from 1 (not 0) if SegmentationType is LABELMAP (Sup 243)

241111:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Ion Range Surface Compensator (CP 2401), Isodose levels for RT Dose visualization (CP 2402)

241105:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Ocular Region Imaged Module in Ophthalmic Tomography (CP 2346)

241003:	libsrc/standard/condn.tpl,module/base.tpl: Restore not requiring SegmentationMacro if TILED_FULL (CP 2331) lost when adding Label Map Segmentation

240915:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Add SOP Class and data elements for Height Map Segmentation and Revised Ophthalmic OCT En Face Image (Sup 240)

240914:	libsrc/standard/condn.tpl,sopcl.tpl,iodcomp,module,strval/base.tpl: Add Label Map Segmentation, and remove related experimental profile (Sup 243)

240913:	libsrc/standard/elmdict/dicom3.tpl: Add data element for Clarify Description related to Range Shifter Settings reusing DICONDE Material Density (CP 2373 FT2)

240912:	libsrc/standard/elmdict/dicom3.tpl: Add data element for Differentiate acquisition techniques for RT 3D CBCT imaging (CP 2220)

240822: libsrc/standard/module/base.tpl:	Correctly Add Ethnic Group Code Sequence (CP 2356)

240728: libsrc/standard/condn.tpl,module/base.tpl: Report error when PositionReferenceIndicator not SLIDE_CORNER for microscopy (000605)

240712:	appsrc/dcfile/dcacqmap.script,dcacqmap.man: Fix invalid character range in egrep for position files (000604), typo in documentation

240623:	libsrc/standard/elmdict/diconde.tpl: Add DICONDE data elements for thermography imaging

240612:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Table Top Position Alignment UID to Table Top positions (CP 2376)

240612:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Improve RT equipment mapping for imaging (CP 2374)

240611:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl:	Add Ethnic Group Code Sequence (CP 2356)

240610:	libsrc/standard/elmdict/dicom3.tpl:	Add data element for updated RT Patient Setup Module (CP 2344)

240515:	libsrc/standard/condn.tpl,module/base.tpl: Check ICCProfile and ColorSpace not present in top level dataset Image Pixel Module if Optical Path Module present (000603)

240506:	libsrc/standard/module/vl.tpl: Check AnnotationGroupNumber is not zero (000602)

240324:	libsrc/standard/condn.tpl,iodcomp/base.tpl: Detect FrameOfReference Module if FoR UID present even if position and orientation not, since MBPO in Secondary Capture IOD (CP 2330)

240402:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check AnnotationGroupUID is not duplicated (000601)

240329:	libsrc/standard/elmdict/other.tpl: 3DHISTECH private data elements from conformance statement

240329:	appsrc/dcfile/dcentvfy.cc,libsrc/src/locale/mesgtext.tpl: Check all ReferencedSOPInstanceUID instances are in set of supplied files (000600)

240328:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check sanity of Microscopy Bulk Simple Annotation indices and coordinates (000599)

240328:	libsrc/include/dctool/attr.h,attrtype.h,attrtypo.h,src/dctool/attrmxrd.cc,attrtypo.cc: Add support of in-memory (small) OL VR data elements to get values to check (000599)

240325:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Clinical Trial Identifiers need Issuers and may be multiple (CP2335)

240324:	libsrc/standard/condn.tpl,iodcomp,module/base.tpl: Clarify Segment scan order for TILED_FULL (CP 2331) and remove CP2331 profile option to test TILED_FULL segmentations

240324:	libsrc/standard/condn.tpl,iodcomp/base.tpl: Add Spatial Information to Secondary Capture IOD (CP 2330)

240324:	libsrc/standard/elmdict/dicom3.tpl,module/vl.tpl: Add information about Vertex Distance to refraction correction parameters (CP 2324)

240324:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Add Synthetic Data Flag (CP 2320)

240316:	appsrc/dcfile/dcentvfy.cc: Check Frame of Reference IE attribute values same for same Series (000597)

240312:	libsrc/standard/elmdict/philips.tpl: Add more observed private data elements

240224:	appsrc/dcfile/dciodvfy.cc,libsrc/include/dctool/attr.h,attrothr.h,transyn.h,src/dctool/transyn.cc,transyne.h,locale/mesgtext.tpl,standard/transyn.tpl,support/transyn.awk: Add check that Basic Offset Table is valid if present, including length, point to Item Tag and begin with a JPEG/J2K SOI marker (000531) and point to even offsets (000526)

240222:	appsrc/acrnema/andump.cc,libsrc/include/dctool/attr.h,attrtype.h,attrtypn.h,attrtypo.h,dcstream.h,vr.h,generic/bnstream.h,src/dctool/attr.cc,attrmxrd.cc,attrnew.cc,attrtypo.cc: Add 64 bit binary VRs (CP 1819)

240215:	libsrc/standard/condn.tpl,module/vl.tpl: Correct spurious report of invalid CommonZCoordinateValue for nnotationCoordinateType not 3D when not actually present (000596)

240214:	appsrc/dcfile/dcentvfy.cc: Check Equipment IE attribute values same for same Series (000595)

240131:	libsrc/standard/condn.tpl,iodcomp,module,strval/base.tpl: Add profile option to test LABELMAP segmentations (Sup243)

240118:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check that SCOORD3D POLYGON are closed (first and last points same) (000592)

240114:	libsrc/standard/elmdict/dicom3.tpl: Provide additional ROI parameters (CP 2296)

240112:	libsrc/standard/elmdict/dicom3.tpl: Add new data elements for DICOMweb Storage Commitment Service (Sup 234)

231125:	config/Configure: Build two architectures when clang >=15 but not with ld flag "-force_cpusubtype_ALL" that is no longer supported

231115:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Add Confocal Microscopy data elements and SOP Class UIDs (Sup 236)


appsrc/dcfile/dccp.cc
static const char *CopyrightIdentifier(void) { return "@(#)dccp.cc Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved."; }
#include "attrmxls.h"
#include "attrseq.h"
#include "mesgtext.h"
#include "dcopt.h"
#include "elmconst.h"

void
keepOnlySelectedLUTSequenceItem(AttributeList &list,Tag lutSequenceTag,int whichLUTItemToKeep,TextOutputStream &log) {
	Attribute *aLUTSequence = list[lutSequenceTag];
	if (aLUTSequence) {
		AttributeList **items;
		int nLUTItems=aLUTSequence->getLists(&items);
		if (whichLUTItemToKeep >= 0 && whichLUTItemToKeep < nLUTItems) {
			AttributeList *itemToKeep = items[whichLUTItemToKeep];
			SequenceAttribute *newLUTSequence = new SequenceAttribute(lutSequenceTag);
			Assert(newLUTSequence);
			(*newLUTSequence)+=itemToKeep;
			list-=lutSequenceTag;
			list+=newLUTSequence;
		}
		else {
			log << WMsgDC("Missing LUT Sequence Item") << " - " << lutSequenceTag << " # " << whichLUTItemToKeep << endl;
		}
	}
	else {
		log << WMsgDC("Missing LUT Sequence") << " - " << lutSequenceTag << endl;
	}
}

// writeTagNumberAndNameToLog is copied from appsrc/dcfile/dciodvfy.cc
// Similar to TextOutputStream& Tag::write(TextOutputStream& stream,ElementDictionary *dict) in libsrc/src/dctool/attrtag.cc
// but without the VR

static void writeTagNumberAndNameToLog(Tag tag,ElementDictionary *dict,TextOutputStream &log) {
	log << "(";
	writeZeroPaddedHexNumber(log,tag.getGroup(),4);
	log << ",";
	writeZeroPaddedHexNumber(log,tag.getElement(),4);
	log << ")";
	const char *desc = NULL;
	if (dict) {
		desc = dict->getDescription(tag);
	}
	if (desc && strlen(desc) > 0 && strcmp(desc,"?") != 0) {
		log << " " << desc;
	}
}

static void writeTagNumberAndNameToLog(Attribute *a,ElementDictionary *dict,TextOutputStream &log) {
	writeTagNumberAndNameToLog(a->getTag(),dict,log);
}

// loopOverListsInSequencesWithLog is copied from libsrc/src/dctool/attrmxls.cc

static bool
loopOverListsInSequencesWithLog(Attribute *a,TextOutputStream &log,
		bool (*func)(AttributeList &,TextOutputStream &))
{
	bool succeeded=true;
	if (strcmp(a->getVR(),"SQ") == 0) {
		AttributeList **al;
		int n;
		if ((n=a->getLists(&al)) > 0) {
			int i;
			for (i=0; i<n; ++i) {
				if (!(*func)(*al[i],log)) succeeded=false;
			}
			delete [] al;
		}
	}
	return succeeded;
}

static bool
fixBadDecimalSeparator(AttributeList &list,TextOutputStream &log) {
	bool success=true;
	AttributeListIterator listi(list);
	while (!listi) {
		Attribute *a=listi();
		Assert(a);
		if (strcmp(a->getVR(),"DS") == 0) {
			int vm = a->getVM();
			int i;
			for (i=0; i<vm; ++i) {
				char *value;
				a->getValue(i,value);
				bool changedsomething=false;
				char *p = value;
				while (*p != 0) {
					if (*p == ',') {
						*p='.';
						changedsomething=true;
					}
					++p;
				}
				if (changedsomething) {
					log << WMsgDC(FixingBadDecimalSeparator) << " in DS VR - ";
					writeTagNumberAndNameToLog(a,list.getDictionary(),log);
					log << " - new value " << (i+1) << " is \"" << value << "\"" << endl;
					a->setValue(i,value);
				}
			}
		}
		if (!::loopOverListsInSequencesWithLog(a,log,&::fixBadDecimalSeparator)) {
			success=false;
		}
		++listi;
	}
	return success;
}

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);
	DicomOutputOptions 	dicom_output_options(options);

	bool verbose=options.get("verbose") || options.get("v");
	bool ignorereaderrors=options.get("ignorereaderrors");
	bool removecommandgroup=options.get("removecommandgroup");
	bool fixbaddecimalseparator=options.get("fixbaddecimalseparator");

	int whichVOILUTitem=0;
	bool selectVOILUTitem=options.get("selectvoilutitem",whichVOILUTitem);

	dicom_input_options.done();
	dicom_output_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);
	DicomOutputOpenerFromOptions output_opener(
		options,dicom_output_options.filename,cout);

	cerr << dicom_input_options.errors();
	cerr << dicom_output_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();
	cerr << output_opener.errors();

	if (!dicom_input_options.good()
	 || !dicom_output_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !output_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< dicom_output_options.usage()
			<< " [-v|-verbose]"
			<< " [-ignorereaderrors]"
			<< " [-removecommandgroup]"
			<< " [-selectvoilutitem 0..n-1]"
			<< " [-fixbaddecimalseparator]"
			<< " [" << MMsgDC(InputFile)
				<< "[" << MMsgDC(OutputFile) << "]]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);
	DicomOutputStream dout(*(ostream *)output_opener,
		dicom_output_options.transfersyntaxuid,
		dicom_output_options.usemetaheader,
		dicom_output_options.useimplicitmetaheader,
		dicom_output_options.addtiff);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream  log(cerr);
	if (verbose) log << "******** While reading ... ********" << endl; 
	list.read(din,false/*newformat*/,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	if (!list.good()) {
		if (!ignorereaderrors) {
			log << list.errors();
			success=false;
		}
		log << EMsgDC(DatasetReadFailed) << endl;
	}

	if (success) {
		// Don't write list warnings yet ...
		// done in usualManagedAttributeListWrite ...
		
		if (removecommandgroup) {
			list.removeCommandGroup();
		}
		
		if (selectVOILUTitem) {
			keepOnlySelectedLUTSequenceItem(list,TagFromName(VOILUTSequence),whichVOILUTitem,log);
		}
		
		if (fixbaddecimalseparator) {
			fixBadDecimalSeparator(list,log);
		}

		if (!usualManagedAttributeListWrite(list,dout,
			dicom_output_options,log,verbose)) success=false;
	}

	return success ? 0 : 1;
}
